lib/rucio/core/transfer.py

```diff
@@ -1063,6 +1064,88 @@ def apply(self, ctx: RequestRankingContext, source: RequestSource) -> "Optional[
         return path[0].src.distance
 
 
+class TransferWaitTime(SourceRankingStrategy):
+    """
+    A source ranking strategy that ranks source nodes based on the expected wait time in queue with exploration embedded
+    through randomization. Note that smaller expected wait times are better.
+    """
+    # There is ~10 s of overhead for any file transfer
+    FILE_OVERHEAD_MS = 10000
+    # Assume 10 Mbps transfer speed
+    PER_BYTE_MS = 1000 / ((10 ** 7) / 8)
+    # Probability we choose an above-average transfer time source node
+    ABOVE_AVG_EXPLORATION = 0.1
+
+    class _RankingContext(RequestRankingContext):
+        def __init__(
+                self,
+                strategy: "TransferWaitTime",
+                rws: "RequestWithSources",
+                costs: "Mapping[RseData, int]"
+        ) -> None:
+            super().__init__(strategy, rws)
+            self.costs = costs
+
+    def for_request(
+            self,
+            rws: RequestWithSources,
+            sources: "Iterable[RequestSource]",
+            *,
+            logger: "LoggerFunction" = logging.log,
+            session: "Session"
+    ) -> "RequestRankingContext":
+        wait_by_rse = {}
+
+        # For each source, estimate the average queue waiting time
+        for src in sources:
+            src_files_queued = 0
+            src_bytes_queued = 0
+            metrics = request_core.get_request_metrics(src_rse_id=src.rse.id, session=session)
```

Comment:
This call to get_request_metrics hides a very costly operation. Especially if done in a loop. The database table is not indexed by src_rse_id. So it will result in a full table scan. Currently we are approaching a million rows in it.

Reply:
It still isn't indexed by the request source as of today. But we were discussing about adding that index for a long time. So may be worth doing it.